cobalt/akn.py

```diff
@@ -0,0 +1,403 @@
+        """
+        root = objectify.fromstring(xml, parser=self._parser)
+
+        # ensure the root element is correct
+        name = root.tag.split('}', 1)[1]
+        if name != 'akomaNtoso':
+            raise ValueError(f"XML root element must be akomaNtoso, but got {name} instead")
+
+        return root
+
+    def to_xml(self, encoding='utf-8'):
+        return etree.tostring(self.root, encoding=encoding)
+
+    def get_namespace(self):
+        akn_namespaces = [ns[1] for ns in sorted(list(AKN_NAMESPACES.items()), reverse=True)]
+        namespaces = list(self.root.nsmap.values())
+        for ns in akn_namespaces:
+            if ns in namespaces:
+                return ns
+
+        raise ValueError(f"Expected to find one of the following Akoma Ntoso XML namespaces: {', '.join(akn_namespaces)}. Only these namespaces were found: {', '.join(namespaces)}")
+
+
+class Fragment(AkomaNtosoDocument):
+    pass
+
+
+# ----------------------------------------------------------------------------------
+# Document structure classes
+
+class StructuredDocument(AkomaNtosoDocument):
+    """ Common base class for AKN documents with a known document structure.
+    """
+
+    structure_type = None
+    """ The name of the this document's structural type.
+    """
+
+    main_content_element = None
+    """ The name of the structural type's main content element.
+    """
+
+    document_type = None
+    """ The name of the document type, corresponding to the primary document XML element.
+    """
+
+    def __init__(self, xml=None):
+        """ Setup a new instance with the string in `xml`. """
+        if not xml:
+            # use an empty document
+            xml = self.EMPTY_DOCUMENT
+        super().__init__(xml)
+
+        # make, eg. ".act" an alias for ".main"
+        setattr(self, self.document_type, self.main)
+
+        # make, eg. ".body" an alias for ".main_content"
+        setattr(self, self.main_content_element, self.main_content)
+
+    def parse(self, xml, document_type=None):
+        """ Parse XML and ensure it's Akoma Ntoso.
+        Raises ValueError on error. Returns the root element.
+        """
+        root = super().parse(xml, document_type)
+
+        if root.countchildren() < 1:
+            raise ValueError("XML root element must have at least one child")
+
+        name = root.getchildren()[0].tag.split('}', 1)[1]
+        if name != self.document_type:
+            raise ValueError(f"Expected {self.document_type} as first child of root element, but got {name} instead")
+
+        return root
+
+    @property
+    def main(self):
+        """ Get the root document element.
+        """
+        return getattr(self.root, self.document_type)
+
+    @property
+    def main_content(self):
+        """ Get the main content element of the document.
+        """
+        return getattr(self.main, self.main_content_element)
```

Comment:
In retrospect I think it's confusing that `main_content_element` isn't an XML element. Instead, let's rename `main_content_element` to `main_content_tag` (and any other variables that should be `_tag` instead of `_element`). Then it's clearer that it's a tag name.